File: typescript/packages/mechanisms/evm/test/unit/batch-settlement/voucherStore.test.ts
import { describe, it, expect } from "vitest";
import {
  channelStateExtra,
  commitVoucherCharge,
  isFacilitatorManaged,
  paymentResponseExtra,
  pendingTtlMs,
  voucherStoreMode,
} from "../../../src/batch-settlement/voucherStore";
import {
  InMemoryChannelStorage,
  type Channel,
  type ChannelStorage,
} from "../../../src/batch-settlement/storage/channel";
import type { PaymentRequirements } from "@x402/core/types";

const CHANNEL_ID = "0xabc1230000000000000000000000000000000000000000000000000000000001";

function baseChannel(overrides: Partial<Channel> = {}): Channel {
  return {
    channelId: CHANNEL_ID,
    channelConfig: {
      payer: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
      payerAuthorizer: "0x0000000000000000000000000000000000000000",
      receiver: "0x9876543210987654321098765432109876543210",
      receiverAuthorizer: "0x1111111111111111111111111111111111111111",
      token: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
      withdrawDelay: 900,
      salt: "0x0000000000000000000000000000000000000000000000000000000000000000",
    },
    chargedCumulativeAmount: "2000",
    signedMaxClaimable: "3000",
    signature: "0xaaa",
    balance: "10000",
    totalClaimed: "0",
    withdrawRequestedAt: 0,
    refundNonce: 0,
    lastRequestTimestamp: 1,
    ...overrides,
  };
}

describe("voucherStore helpers — pendingTtlMs", () => {
  it("clamps a zero timeout to the minimum admission lock TTL", () => {
    expect(pendingTtlMs(0)).toBe(5_000);
  });

  it("clamps an oversized timeout to ten minutes", () => {
    expect(pendingTtlMs(3600)).toBe(600_000);
  });

  it("honours a mid-range timeout in milliseconds", () => {
    expect(pendingTtlMs(120)).toBe(120_000);
  });

  it("treats an undefined timeout like zero before clamping", () => {
    expect(pendingTtlMs(undefined)).toBe(5_000);
  });
});

describe("voucherStore helpers — mode detection", () => {
  it("treats only explicit true as facilitator-managed", () => {
    expect(isFacilitatorManaged({ extra: { voucherStore: true } })).toBe(true);
    expect(isFacilitatorManaged({ extra: { voucherStore: false } })).toBe(false);
    expect(isFacilitatorManaged({ extra: {} })).toBe(false);
    expect(voucherStoreMode({ extra: { voucherStore: true } } as PaymentRequirements)).toBe(
      "facilitator",
    );
    expect(voucherStoreMode({ extra: {} } as PaymentRequirements)).toBe("self");
  });
});

describe("voucherStore helpers — commitVoucherCharge", () => {
  it("returns missing when there is no row and no snapshot", async () => {
    const storage = new InMemoryChannelStorage();
    const result = await commitVoucherCharge(storage, CHANNEL_ID, {
      increment: 1000n,
      signedCap: 5000n,
      voucher: { maxClaimableAmount: "5000", signature: "0xbbb" },
    });
    expect(result).toEqual({ status: "missing" });
  });

  it("returns cap_exceeded without mutating storage", async () => {
    const storage = new InMemoryChannelStorage();
    await storage.updateChannel(CHANNEL_ID, () => baseChannel());
    const result = await commitVoucherCharge(storage, CHANNEL_ID, {
      increment: 2000n,
      signedCap: 3000n,
      voucher: { maxClaimableAmount: "3000", signature: "0xbbb" },
    });
    expect(result).toEqual({ status: "cap_exceeded", charged: "4000" });
    expect(await storage.get(CHANNEL_ID)).toMatchObject({ chargedCumulativeAmount: "2000" });
  });

  it("applies map after a successful commit", async () => {
    const storage = new InMemoryChannelStorage();
    await storage.updateChannel(CHANNEL_ID, () => baseChannel());
    const result = await commitVoucherCharge(storage, CHANNEL_ID, {
      increment: 500n,
      signedCap: 3000n,
      voucher: { maxClaimableAmount: "2500", signature: "0xccc" },
      map: channel => ({ ...channel, chargedCumulativeAmount: "9999" }),
    });
    expect(result.status).toBe("committed");
    if (result.status === "committed") {
      expect(result.current.chargedCumulativeAmount).toBe("9999");
    }
  });

  it("creates a channel row from snapshot when storage is empty", async () => {
    const storage = new InMemoryChannelStorage();
    const snapshot = baseChannel({ chargedCumulativeAmount: "1000", balance: "9000" });
    const result = await commitVoucherCharge(storage, CHANNEL_ID, {
      increment: 500n,
      signedCap: 5000n,
      voucher: { maxClaimableAmount: "5000", signature: "0xbbb" },
      snapshot,
    });
    expect(result.status).toBe("committed");
    if (result.status === "committed") {
      expect(result.current.chargedCumulativeAmount).toBe("1500");
      expect(result.current.balance).toBe("9000");
    }
    expect(await storage.get(CHANNEL_ID)).toBeDefined();
  });

  it("returns conflict when storage does not apply the charge update", async () => {
    const storage: ChannelStorage = {
      get: async () => undefined,
      list: async () => [],
      updateChannel: async () => ({ channel: baseChannel(), status: "unchanged" }),
    };
    const result = await commitVoucherCharge(storage, CHANNEL_ID, {
      increment: 500n,
      signedCap: 5000n,
      voucher: { maxClaimableAmount: "5000", signature: "0xbbb" },
      snapshot: baseChannel(),
    });
    expect(result).toEqual({ status: "conflict" });
  });

  it("keeps stored escrow fields when localVerify is true even if a snapshot is provided", async () => {
    const storage = new InMemoryChannelStorage();
    await storage.updateChannel(CHANNEL_ID, () =>
      baseChannel({ balance: "7777", totalClaimed: "3" }),
    );
    const snapshot = baseChannel({
      balance: "10000",
      totalClaimed: "0",
    });
    const result = await commitVoucherCharge(storage, CHANNEL_ID, {
      increment: 100n,
      signedCap: 5000n,
      voucher: { maxClaimableAmount: "5000", signature: "0xbbb" },
      snapshot,
      localVerify: true,
    });
    expect(result.status).toBe("committed");
    if (result.status === "committed") {
      expect(result.current.balance).toBe("7777");
      expect(result.current.totalClaimed).toBe("3");
    }
  });

  it("refreshes escrow fields from a facilitator snapshot when not in local verify mode", async () => {
    const storage = new InMemoryChannelStorage();
    await storage.updateChannel(CHANNEL_ID, () =>
      baseChannel({ balance: "1", totalClaimed: "9", refundNonce: 1, withdrawRequestedAt: 2 }),
    );
    const snapshot = baseChannel({
      balance: "10000",
      totalClaimed: "0",
      refundNonce: 0,
      withdrawRequestedAt: 0,
    });
    const result = await commitVoucherCharge(storage, CHANNEL_ID, {
      increment: 1000n,
      signedCap: 5000n,
      voucher: { maxClaimableAmount: "5000", signature: "0xbbb" },
      snapshot,
    });
    expect(result.status).toBe("committed");
    if (result.status === "committed") {
      expect(result.current.balance).toBe("10000");
      expect(result.current.totalClaimed).toBe("0");
      expect(result.current.chargedCumulativeAmount).toBe("3000");
    }
  });
});

describe("voucherStore helpers — response extras", () => {
  it("omits chargedCumulativeAmount from channelStateExtra when not provided", () => {
    expect(channelStateExtra(baseChannel())).not.toHaveProperty("chargedCumulativeAmount");
  });

  it("includes chargeCount only for facilitator-managed paid responses", () => {
    const snapshot = channelStateExtra(baseChannel(), "2500");
    const channelOnly = paymentResponseExtra({ channelState: snapshot });
    expect(Object.keys(channelOnly)).toEqual(["channelState"]);
    const paid = paymentResponseExtra({
      channelState: snapshot,
      chargedAmount: "500",
      chargeCount: 2,
    });
    expect(Object.keys(paid)).toEqual(["channelState", "chargedAmount", "chargeCount"]);
    const refund = paymentResponseExtra({ channelState: snapshot, chargeCount: 0 });
    expect(refund.chargedAmount).toBeUndefined();
    expect(refund.chargeCount).toBe(0);
  });
});

File: go/mechanisms/evm/batch-settlement/storage/voucher_store_test.go
package storage

import (
	"context"
	"math/big"
	"testing"

	batchsettlement "github.com/x402-foundation/x402/go/v2/mechanisms/evm/batch-settlement"
	"github.com/x402-foundation/x402/go/v2/types"
)

const voucherChannelId = "0xabc1230000000000000000000000000000000000000000000000000000000001"

func voucherBaseChannel(overrides *Channel) *Channel {
	out := &Channel{
		ChannelId: voucherChannelId,
		ChannelConfig: batchsettlement.ChannelConfig{
			Payer:              "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
			PayerAuthorizer:    "0x0000000000000000000000000000000000000000",
			Receiver:           "0x9876543210987654321098765432109876543210",
			ReceiverAuthorizer: "0x1111111111111111111111111111111111111111",
			Token:              "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
			WithdrawDelay:      900,
			Salt:               "0x0000000000000000000000000000000000000000000000000000000000000000",
		},
		ChargedCumulativeAmount: "2000",
		SignedMaxClaimable:      "3000",
		Signature:               "0xaaa",
		Balance:                 "10000",
		TotalClaimed:            "0",
		WithdrawRequestedAt:     0,
		RefundNonce:             0,
		LastRequestTimestamp:    1,
	}
	if overrides == nil {
		return out
	}
	if overrides.ChargedCumulativeAmount != "" {
		out.ChargedCumulativeAmount = overrides.ChargedCumulativeAmount
	}
	if overrides.Balance != "" {
		out.Balance = overrides.Balance
	}
	if overrides.TotalClaimed != "" {
		out.TotalClaimed = overrides.TotalClaimed
	}
	if overrides.RefundNonce != 0 {
		out.RefundNonce = overrides.RefundNonce
	}
	if overrides.WithdrawRequestedAt != 0 {
		out.WithdrawRequestedAt = overrides.WithdrawRequestedAt
	}
	return out
}

func TestPendingTtlMs_ClampsZeroToMinimum(t *testing.T) {
	if got := PendingTtlMs(0); got != 5_000 {
		t.Fatalf("got %d", got)
	}
}

func TestPendingTtlMs_ClampsOversizedToTenMinutes(t *testing.T) {
	if got := PendingTtlMs(3600); got != 600_000 {
		t.Fatalf("got %d", got)
	}
}

func TestPendingTtlMs_HonoursMidRange(t *testing.T) {
	if got := PendingTtlMs(120); got != 120_000 {
		t.Fatalf("got %d", got)
	}
}

func TestDefaultOnchainStateTtlMs_ClampsShortDelay(t *testing.T) {
	if got := DefaultOnchainStateTtlMs(30); got != 30_000 {
		t.Fatalf("got %d", got)
	}
}

func TestDefaultOnchainStateTtlMs_UsesOneThird(t *testing.T) {
	if got := DefaultOnchainStateTtlMs(180); got != 60_000 {
		t.Fatalf("got %d", got)
	}
}

func TestDefaultOnchainStateTtlMs_ClampsLongDelay(t *testing.T) {
	if got := DefaultOnchainStateTtlMs(900); got != 300_000 {
		t.Fatalf("got %d", got)
	}
}

func TestAdmissionOwner_StableForSameInputs(t *testing.T) {
	voucher := batchsettlement.BatchSettlementVoucherFields{
		ChannelId:          voucherChannelId,
		MaxClaimableAmount: "1000",
		Signature:          "0xfeedface",
	}
	first := AdmissionOwner("0xpending", voucher)
	second := AdmissionOwner("0xpending", voucher)
	if first != second {
		t.Fatal("expected stable owner")
	}
}

func TestAdmissionOwner_ChangesWithVoucher(t *testing.T) {
	voucher := batchsettlement.BatchSettlementVoucherFields{
		ChannelId:          voucherChannelId,
		MaxClaimableAmount: "1000",
		Signature:          "0xfeedface",
	}
	owner := AdmissionOwner("0xpending", voucher)
	changedSig := voucher
	changedSig.Signature = "0xdeadbeef"
	changedCap := voucher
	changedCap.MaxClaimableAmount = "2000"
	if AdmissionOwner("0xpending", changedSig) == owner {
		t.Fatal("signature change should change owner")
	}
	if AdmissionOwner("0xpending", changedCap) == owner {
		t.Fatal("cap change should change owner")
	}
}

func TestAdmissionOwner_NormalizesLowercase(t *testing.T) {
	voucher := batchsettlement.BatchSettlementVoucherFields{
		ChannelId:          voucherChannelId,
		MaxClaimableAmount: "1000",
		Signature:          "0xfeedface",
	}
	upper := voucher
	upper.Signature = "0xFEEDFACE"
	if AdmissionOwner("0xPENDING", upper) != AdmissionOwner("0xpending", voucher) {
		t.Fatal("owner should be case-insensitive")
	}
}

func TestIsFacilitatorManaged_OnlyExplicitTrue(t *testing.T) {
	if !IsFacilitatorManaged(map[string]interface{}{"voucherStore": true}) {
		t.Fatal("true should be managed")
	}
	if IsFacilitatorManaged(map[string]interface{}{"voucherStore": false}) {
		t.Fatal("false should not be managed")
	}
	if IsFacilitatorManaged(map[string]interface{}{}) {
		t.Fatal("empty extra should not be managed")
	}
	if VoucherStoreModeOf(types.PaymentRequirements{Extra: map[string]interface{}{"voucherStore": true}}) != VoucherStoreModeFacilitator {
		t.Fatal("expected facilitator mode")
	}
	if VoucherStoreModeOf(types.PaymentRequirements{Extra: map[string]interface{}{}}) != VoucherStoreModeSelf {
		t.Fatal("expected self mode")
	}
}

func TestCommitVoucherCharge_MissingWithoutSnapshot(t *testing.T) {
	store := NewInMemoryChannelStorage[*Channel]()
	result, err := CommitVoucherCharge(context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment: big.NewInt(1000),
		SignedCap: big.NewInt(5000),
		Voucher:   batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "5000", Signature: "0xbbb"},
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitMissing {
		t.Fatalf("status = %q", result.Status)
	}
}

func TestCommitVoucherCharge_CapExceededDoesNotMutate(t *testing.T) {
	store := NewInMemoryChannelStorage[*Channel]()
	if _, err := store.UpdateChannel(context.Background(), voucherChannelId, func(*Channel) *Channel { return voucherBaseChannel(nil) }); err != nil {
		t.Fatalf("seed: %v", err)
	}
	result, err := CommitVoucherCharge(context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment: big.NewInt(2000),
		SignedCap: big.NewInt(3000),
		Voucher:   batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "3000", Signature: "0xbbb"},
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitCapExceeded || result.Charged != "4000" {
		t.Fatalf("result = %+v", result)
	}
	got, _ := store.Get(context.Background(), voucherChannelId)
	if got.ChargedCumulativeAmount != "2000" {
		t.Fatalf("charged = %q", got.ChargedCumulativeAmount)
	}
}

func TestCommitVoucherCharge_CorruptWatermarkIsConflictWithoutWrite(t *testing.T) {
	store := NewInMemoryChannelStorage[*Channel]()
	seed := voucherBaseChannel(nil)
	seed.ChargedCumulativeAmount = "not-a-number"
	if _, err := store.UpdateChannel(context.Background(), voucherChannelId, func(*Channel) *Channel { return seed }); err != nil {
		t.Fatalf("seed: %v", err)
	}
	result, err := CommitVoucherCharge(context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment: big.NewInt(100),
		SignedCap: big.NewInt(100),
		Voucher:   batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "100", Signature: "0xbbb"},
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitConflict {
		t.Fatalf("status = %q, want conflict", result.Status)
	}
	got, _ := store.Get(context.Background(), voucherChannelId)
	if got.ChargedCumulativeAmount != "not-a-number" {
		t.Fatalf("corrupt watermark was overwritten: %q", got.ChargedCumulativeAmount)
	}
}

func TestCommitVoucherCharge_AppliesMapAfterCommit(t *testing.T) {
	store := NewInMemoryChannelStorage[*Channel]()
	if _, err := store.UpdateChannel(context.Background(), voucherChannelId, func(*Channel) *Channel { return voucherBaseChannel(nil) }); err != nil {
		t.Fatalf("seed: %v", err)
	}
	result, err := CommitVoucherCharge(context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment: big.NewInt(500),
		SignedCap: big.NewInt(3000),
		Voucher:   batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "2500", Signature: "0xccc"},
		Map: func(channel *Channel) *Channel {
			cp := *channel
			cp.ChargedCumulativeAmount = "9999"
			return &cp
		},
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitCommitted || result.Current.ChargedCumulativeAmount != "9999" {
		t.Fatalf("result = %+v", result)
	}
}

func TestCommitVoucherCharge_CreatesFromSnapshot(t *testing.T) {
	store := NewInMemoryChannelStorage[*Channel]()
	snapshot := voucherBaseChannel(&Channel{ChargedCumulativeAmount: "1000", Balance: "9000"})
	result, err := CommitVoucherCharge(context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment: big.NewInt(500),
		SignedCap: big.NewInt(5000),
		Voucher:   batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "5000", Signature: "0xbbb"},
		Snapshot:  snapshot,
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitCommitted || result.Current.ChargedCumulativeAmount != "1500" || result.Current.Balance != "9000" {
		t.Fatalf("result = %+v", result)
	}
	got, _ := store.Get(context.Background(), voucherChannelId)
	if got == nil {
		t.Fatal("expected stored row")
	}
}

type unchangedChargeStore struct {
	*InMemoryChannelStorage[*Channel]
}

func (s unchangedChargeStore) UpdateChannel(context.Context, string, func(*Channel) *Channel) (*ChannelUpdateResult[*Channel], error) {
	return &ChannelUpdateResult[*Channel]{Channel: voucherBaseChannel(nil), Status: ChannelUnchanged}, nil
}

func TestCommitVoucherCharge_ConflictWhenUpdateNotApplied(t *testing.T) {
	store := unchangedChargeStore{InMemoryChannelStorage: NewInMemoryChannelStorage[*Channel]()}
	result, err := CommitVoucherCharge[*Channel](context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment: big.NewInt(500),
		SignedCap: big.NewInt(5000),
		Voucher:   batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "5000", Signature: "0xbbb"},
		Snapshot:  voucherBaseChannel(nil),
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitConflict {
		t.Fatalf("status = %q", result.Status)
	}
}

func TestCommitVoucherCharge_LocalVerifyKeepsStoredEscrow(t *testing.T) {
	store := NewInMemoryChannelStorage[*Channel]()
	if _, err := store.UpdateChannel(context.Background(), voucherChannelId, func(*Channel) *Channel {
		return voucherBaseChannel(&Channel{Balance: "7777", TotalClaimed: "3"})
	}); err != nil {
		t.Fatalf("seed: %v", err)
	}
	result, err := CommitVoucherCharge(context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment:   big.NewInt(100),
		SignedCap:   big.NewInt(5000),
		Voucher:     batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "5000", Signature: "0xbbb"},
		Snapshot:    voucherBaseChannel(&Channel{Balance: "10000", TotalClaimed: "0"}),
		LocalVerify: true,
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitCommitted || result.Current.Balance != "7777" || result.Current.TotalClaimed != "3" {
		t.Fatalf("result = %+v", result)
	}
}

func TestCommitVoucherCharge_RefreshesEscrowFromSnapshot(t *testing.T) {
	store := NewInMemoryChannelStorage[*Channel]()
	if _, err := store.UpdateChannel(context.Background(), voucherChannelId, func(*Channel) *Channel {
		return voucherBaseChannel(&Channel{Balance: "1", TotalClaimed: "9", RefundNonce: 1, WithdrawRequestedAt: 2})
	}); err != nil {
		t.Fatalf("seed: %v", err)
	}
	result, err := CommitVoucherCharge(context.Background(), store, voucherChannelId, CommitVoucherChargeInput[*Channel]{
		Increment: big.NewInt(1000),
		SignedCap: big.NewInt(5000),
		Voucher:   batchsettlement.BatchSettlementVoucherFields{MaxClaimableAmount: "5000", Signature: "0xbbb"},
		Snapshot:  voucherBaseChannel(&Channel{Balance: "10000", TotalClaimed: "0", RefundNonce: 0, WithdrawRequestedAt: 0}),
	})
	if err != nil {
		t.Fatalf("CommitVoucherCharge: %v", err)
	}
	if result.Status != CommitCommitted {
		t.Fatalf("status = %q", result.Status)
	}
	if result.Current.Balance != "10000" || result.Current.TotalClaimed != "0" || result.Current.ChargedCumulativeAmount != "3000" {
		t.Fatalf("current = %+v", result.Current)
	}
}

func TestChannelStateExtra_OmitsChargedWhenMissing(t *testing.T) {
	extra := ChannelStateExtra(voucherBaseChannel(nil), nil)
	if extra.ChargedCumulativeAmount != "" {
		t.Fatalf("chargedCumulativeAmount = %q", extra.ChargedCumulativeAmount)
	}
}

func TestPaymentResponseExtra_IncludesChargeCountOnlyWhenSet(t *testing.T) {
	snapshot := ChannelStateExtra(voucherBaseChannel(nil), strPtr("2500"))
	channelOnly := PaymentResponseExtra(snapshot, nil, nil)
	if channelOnly.ChargedAmount != "" || channelOnly.ChargeCount != nil || channelOnly.ChannelState == nil {
		t.Fatalf("channelOnly = %+v", channelOnly)
	}
	count := 2
	charged := "500"
	paid := PaymentResponseExtra(snapshot, &charged, &count)
	if paid.ChargedAmount != "500" || paid.ChargeCount == nil || *paid.ChargeCount != 2 {
		t.Fatalf("paid = %+v", paid)
	}
	zero := 0
	refund := PaymentResponseExtra(snapshot, nil, &zero)
	if refund.ChargedAmount != "" || refund.ChargeCount == nil || *refund.ChargeCount != 0 {
		t.Fatalf("refund = %+v", refund)
	}
}

func strPtr(s string) *string { return &s }
